### PR TITLE
Port cuda sigmoid to Aten(CUDA)

### PR DESCRIPTION
**Pull request by XiaobingSuper**
@VitalyFedyunin, this PR port cuda sigmoid to Aten: https://github.com/pytorch/pytorch/issues/24624; TH/THC sigmoid code can 't be removed because the sigmoid_backward in THNN/THCUNN rely on it.  I will port sigmoid_backward to Aten next step, incuding CPU and CUDA, which will remove the sigmoid code in TH/THC .

Test script:
```
import timeit

 device = "cuda"
 for n, t in [(10, 100000),(1000, 10000)]:
     print('a.sigmoid() (a.numel() == {}) for {} times'.format(n, t))
     for dtype in ('torch.float', 'torch.double', 'torch.half'):
         print(f'device: {device}, dtype: {dtype}, {t} times', end='\t\t')
         print(timeit.timeit(f'a.sigmoid()\nif "{device}" == "cuda": torch.cuda.synchronize()', setup=f'import torch; a = torch.ones({n}, device="{device}", dtype={dtype})', number=t))
```

Device: **Tesla P40**

Before:
```
a.sigmoid() (a.numel() == 10) for 100000 times
device: cuda, dtype: torch.float, 100000 times          1.2853778750286438
device: cuda, dtype: torch.double, 100000 times         1.2787265420192853
device: cuda, dtype: torch.half, 100000 times           1.2610833930084482
a.sigmoid() (a.numel() == 1000) for 10000 times
device: cuda, dtype: torch.float, 10000 times           0.1274153349804692
device: cuda, dtype: torch.double, 10000 times          0.13953313598176464
device: cuda, dtype: torch.half, 10000 times            0.1265286349807866
```
After:
```
a.sigmoid() (a.numel() == 10) for 100000 times
device: cuda, dtype: torch.float, 100000 times          1.275270765996538
device: cuda, dtype: torch.double, 100000 times         1.285128042974975
device: cuda, dtype: torch.half, 100000 times           1.2761492819990963
a.sigmoid() (a.numel() == 1000) for 10000 times
device: cuda, dtype: torch.float, 10000 times           0.12851508799940348
device: cuda, dtype: torch.double, 10000 times          0.13738596899202093
device: cuda, dtype: torch.half, 10000 times            0.12715664599090815
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/UnaryOpsKernel.cu [hunk 1]
```diff
@@ -99,6 +99,15 @@ void sign_kernel_cuda(TensorIterator& iter){
     }
 }
 
+void sigmoid_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "sigmoid_cuda", [&]() {
+      gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
+         scalar_t one = scalar_t(1);
+         return  one / (one + std::exp(- a));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UnaryOpsKernel.cu]: Avoid using std:: functions in the device code, this can cause ROCm failures (see e.g. #25791 that had to be backed out)

Comment (XiaobingSuper) [reply aten/src/ATen/native/cuda/UnaryOpsKernel.cu]: As comment on https://github.com/pytorch/pytorch/blob/7f89464b2d9b20d728c1d29c33d32ada51ab9937/aten/src/ATen/native/Distributions.h#L7, floor also cause ROCm failures, why std::floor is used in  https://github.com/pytorch/pytorch/blob/af96e0cb5bf83175cb648659ce3b40d84a807c77/aten/src/ATen/native/cuda/UnaryOpsKernel.cu#L43

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by vincentqb** [comment]
@VitalyFedyunin I'm rerunning internal tests.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -2466,15 +2466,15 @@
...
+    CPU: sigmoid_
+    CUDA: sigmoid_
     MkldnnCPU: mkldnn_sigmoid_
 
 - func: sigmoid.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
   supports_named_tensor: True
   dispatch:
-    CPU: _sigmoid_out_cpu
-    CUDA: _sigmoid_out_cuda
+    CPU: sigmoid_out
```
Comment (ifedan) [inline aten/src/ATen/native/native_functions.yaml]: this section can be removed, both names are same

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -2466,15 +2466,15 @@
   supports_named_tensor: True
   variants: function, method
   dispatch:
-    CPU: _sigmoid__cpu
-    CUDA: _sigmoid__cuda
+    CPU: sigmoid_
```
Comment (ifedan) [inline aten/src/ATen/native/native_functions.yaml]: this section can be removed, both names are same

**Comment by ifedan** [comment]
can we remove TH implementation?

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/native_functions.yaml]: removed.

**Comment by XiaobingSuper** [comment]
> can we remove TH implementation?

@ifedan, the TH implementation will removed at next step, because the sigmoid backward depend on it. Thanks！

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, please help review this code, thanks1

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, I rebased the code again for the conflicts, thanks!

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 4b552d1:
* 0/19 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=4b552d1a0d8a27924de65496bbe5add605a232f3).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin
